adr/ADR-13.md

```diff
@@ -0,0 +1,290 @@
++OK
+PUB $JS.API.CONSUMER.MSG.NEXT.bar.dur _INBOX.example 0
++OK
+MSG bar 0 $JS.ACK.bar.dur.1.9808.9808.1626818873482533000.0 4
+helo
+```
+
+Note that even though the inbox used for the request was
+`_INBOX.example`, when the message got delivered the subject was
+rewritten into `bar` which is the subject of the message that is
+persisted in JetStream.
+
+A pull request for the next message, will linger until there is no
+more interest in the subject, a client is disconnected or the batch size is filled.
+Each pull request will increase the `num_awaiting` counter for a consumer of
+inflight pull requests [2].  At most, a consumer can only have 512
+inflight pull requests, though this can be changed when creating the
+consumer with the `max_waiting` option [1]:
+
+```shell
+PUB $JS.API.CONSUMER.INFO.bar.dur _INBOX.uMfJLECClHCs0CLfWF7Rsj.ds2ZxC4o 0
+MSG _INBOX.uMfJLECClHCs0CLfWF7Rsj.ds2ZxC4o 1 601
+{
+    "type": "io.nats.jetstream.api.v1.consumer_info_response",
+    "stream_name": "bar",
+    "name": "dur",
+    "created": "2021-07-20T21:35:11.825973Z",
+    "config": {
+        "durable_name": "dur",
+        "deliver_policy": "all",
+        "ack_policy": "explicit",
+        "ack_wait": 30000000000,
+        "max_deliver": -1,
+        "filter_subject": "bar",
+        "replay_policy": "instant",
+        "max_waiting": 512,          <-- Maximum Inflight Pull/Fetch Requests [1]
+        "max_ack_pending": 20000
+    },
+    "delivered": {
+        "consumer_seq": 11561,
+        "stream_seq": 11560
+    },
+    "ack_floor": {
+        "consumer_seq": 11561,
+        "stream_seq": 11560
+    },
+    "num_ack_pending": 0,
+    "num_redelivered": 0,
+    "num_waiting": 1,                <-- Inflight Pull/Fetch Requests [2]
+    "num_pending": 0,
+    "cluster": {
+        "leader": "NCUOUH6YICRESE73CKTXESMGA4ZN4KTELXPUIE6JCRTL6IF4UWE3B2Z4"
+    }
+}
+```
+
+When making a pull request it is also possible to request more than one message:
+
+```shell
+PUB $JS.API.CONSUMER.MSG.NEXT.bar.dur _INBOX.x7tkDPDLCOEknrfB4RH1V7.OgY4M7 11
+{"batch":5}
+```
+
+Whenever a pull request times out, the count of `num_waiting` will increase for a consumer
+but this will eventually reset once it reaches the max waiting inflight that was configured
+for the pull consumer.
+
+### No Wait Pull Requests
+
+In order to get a response from the server rigth away, a client can
+make a pull request with the `no_wait` option enabled. For example:
+
+```shell
+PUB $JS.API.CONSUMER.MSG.NEXT.bar.dur _INBOX.x7tkDPDLCOEknrfB4RH1V7.OgY4M7 26
+{"batch":1,"no_wait":true}
+```
+
+The result of a no wait pull request is a guaranteed instant response
+from the server that will be either the next message or an error,
+where an error could be a server error such as `503` in case JS
+service is not available. Most commonly, the result of a no wait pull
+request will be a `404` no messages error:
+
+```shell
+HMSG _INBOX.x7tkDPDLCOEknrfB4RH1V7.OgY4M7 2  28 28
+NATS/1.0 404 No Messages
+```
```

Comment:
no, it should then fall back to do a pull request that waits and blocks until there are new messages or there is a timeout, so the same request but without `no_wait`

Reply:
So this is not really an "option" that the user set. This is done internally, always, no? The use of the options no_wait, etc.. got confusing because I thought these were options that the user can pass to the "Fetch" api, but this does not seem to be the case then.
If it is, then I don't understand that an option "no_wait" would lead to... a wait for messages ;-)

Reply:
That's right, there is no need for an exposed `no_wait` option since it is just used internally to 'start the session' of being delivered messages by the server. `No Wait` in this context means 'get any response from the server asap' which is either the batch of messages ready to be delivered or an error.